### PR TITLE
Adds a switch to control cleanup behavior after E2E test

### PR DESCRIPTION
**Pull request by StefanHauth**
# Description
It often is hard to analyze failing E2E tests when the environment they failed in is already cleaned up. This PR adds a switch that allows fine tuning cleanup behavior via an environment variable:

`TEST_ENV_CLEANUP`
* `always` - cleanup after each test
* `never` - never clean up
* `onSuccess` - only cleanup if test was successful

## How can this be tested?
* Prepare an environment to run E2E tests in
* set said environment variable to a value of you choice and run E2E tests
* observe if cleanup has happened the way you expected by checking resources on the test cluster

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: test/kubeobjects/environment/environment.go [hunk 1]
```diff
@@ -3,27 +3,126 @@
...
+	return &Environment{
+		Environment: env.NewWithConfig(envConfig),
+		skipCleanup: getSkipCleanupMode(),
+	}
+}
+
+type TestingT interface {
+	Failed() bool
+	Errorf(format string, args ...interface{})
+}
```
Comment (0sewa0) [inline test/kubeobjects/environment/environment.go:63]: why is this needed?

**Review by 0sewa0 — CHANGES_REQUESTED** [review]
this looks like we are waaay overcomplicating things 😅 

just a question wouldn't the following be enough:
```
var testEnvironment env.Environment

func TestMain(m *testing.M) {
	testEnvironment = environment.Get()
	testEnvironment.BeforeEachTest(namespace.DeleteIfExists(sampleapps.Namespace))
	testEnvironment.BeforeEachTest(namespace.DeleteIfExists(dynakube.Namespace))
	testEnvironment.BeforeEachTest(oneagent.WaitForDaemonSetPodsDeletion())
	testEnvironment.BeforeEachTest(namespace.Recreate(namespace.NewBuilder(dynakube.Namespace).Build()))

	if os.Getenv("CLEANUP") == "true" {
		testEnvironment.AfterEachTest(namespace.DeleteIfExists(sampleapps.Namespace))
		testEnvironment.AfterEachTest(dynakube.DeleteIfExists(dynakube.NewBuilder().WithDefaultObjectMeta().Build()))
		testEnvironment.AfterEachTest(oneagent.WaitForDaemonSetPodsDeletion())
		testEnvironment.AfterEachTest(namespace.Delete(dynakube.Namespace))
	}
	testEnvironment.Run(m)
}
```

you could abstract the `os.Getenv("CLEANUP") == "true" ` part to be nicer, but that is it

the `onSuccess` state does not make sense (except 1 scenario), because:
Scenarios you would like to cover:
1. Run ALL tests in ONE pipeline, pointing to the SAME cluster (this is what is the current plan)
In this case you would like to clean up if:
- The cluster you get in the beginning is brand spanking new, no need to pre-clean
- The test passed, and want to run the next test
- The test failed,  and want to run the next test
- **SUMMERY: So always after a test, no ifs or buts (you could argue about the last test, but it doesn't make sense)**
2. Run EACH tests in DIFFERENT pipelines, pointing to DIFFERENT clusters (this is what is the "old" plan AFAIK)
In this case you would like to clean up if:
- The cluster you get in the beginning is brand spanking new, no need to pre-clean
- **SUMMERY:  Never, if test passed cluster goes bye bye, if not, we dont clean up**
3. Run locally, on your special cluster.
- It depends, here `onSuccess` would be a nice to have, but IMO this complex change does not worth the minor connivance it would bring

A would also question the necessity of the `testEnvironment.BeforeEachTest` calls, as they are not needed in the pipeline, and in a local scenario you CAN'T predict every kind of weird stuff that one might have installed a week or so ago in their environment, so it can easily occur that the pre-clean is not good enough, so it adds very little value.

**Comment by StefanHauth** [comment]
Feature is no longer needed, at least in the intended way. Back to the drawing board
